torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -2526,6 +2526,36 @@ Tensor* TensorExprKernel::convertOutputToCorrectStrides(torch::jit::Value* v) {
       });
 }
 
+void TensorExprKernel::bindConstant(const torch::jit::Value* v) {
+  if (!v->type()->cast<TensorType>()) {
+    // Only Tensor constants need to be bound, scalar constants will be turned
+    // into immediates in TE IR
+    return;
+  }
+  auto n = v->node();
+  auto const_tensor = toIValue(v)->toTensor();
+
+  const auto& tt = v->type()->expect<TensorType>();
+  const auto sizes = *tt->sizes().concrete_sizes();
+  std::vector<ExprHandle> te_sizes;
+  for (auto s : sizes) {
+    te_sizes.push_back(IntImm::make(s));
+  }
+
+  const Buf* buf = new Buf(
+      "const_" + v->debugName(),
+      ExprHandleVectorToExprVector(te_sizes),
+      ToDtype(static_cast<ScalarType>(*tt->scalarType())));
+
+  if (!const_tensor.is_contiguous()) {
+    const_tensor = const_tensor.clone().contiguous();
+    unpacked_constant_tensors_.push_back(const_tensor);
```

Comment:
Something seems fishy here but I don't understand `at::Tensor` that well... if `const_tensor` isn't contiguous we `clone()` it and squirrel away the `data_ptr` in `constants_`... but then doesn't the `const_tensor.clone()` go out of scope, get deleted, and we have a dangling pointer?

Also, why do we need to store `unpacked_constant_tensors_`?  (I feel like these must be related but I don't understand how 😅 )

Reply:
AFAIU, `at::Tensor` is a kind of shared ptr, so it could go out of the scope without bad consequences as long as there is at least one live pointer. The reason we're storing it to `unpacked_constant_tensors_` is exactly to keep it alive - this way the `TensorExprKernel` objects owns all constant tensors it materialized via `contiguous` call.

Reply:
Oh this makes sense now.  I totally didn't read the code correctly, and now I see that you're storing `const_tensor.clone().contiguous()` back into `const_tensor`.  So now it's perfectly clear to me how the reference is kept alive.  Thanks!

Reply:
Yep, `at::Tensor` is just a ref-counted pointer to `TensorImpl`, see https://github.com/pytorch/pytorch/blob/7636cb6bab1c7ac2c0285935793404bef3bb546a/aten/src/ATen/templates/TensorBody.h#L71-L88